Overdrive integration

bpalme opened the pull request:
I thought I'd get this pull request started so you can start looking it over.  I still have some work todo.

- [x] improve staff view tab (arrayview doesn't work with OD format b/c of nested arrays)
- [x] add preview (samples) feature
- [x] improve formats tab layout
- [x] refactoring in Overdriveconnector
- [x] implement downloading content feature
- [x] implement return item feature
- [x] better formatting for My Overdrive content template
- [x] better formatting for hold template
- [x] ~~implement checkout history~~ (next version)
- [x] double check that marc record support still works
- [x] Get Overdrive API Team to approve the Integration.
- [x] Solicit testing for this fork
- [x] Resolve https://vufind.org/jira/browse/VUFIND-933 when merged

demiankatz commented on module/VuFind/config/module.config.php:
```diff
@@ -659,9 +675,10 @@
+    'MyResearch/Verify', 'Overdrive/MyContent',
     'Primo/Advanced', 'Primo/Home', 'Primo/Search',
     'QRCode/Show', 'QRCode/Unavailable',
+    'Overdrive/Hold',
```
Minor nitpick: this has gotten out of alphabetical order.

demiankatz commented on module/VuFind/src/VuFind/Controller/AbstractBase.php:
```diff
@@ -136,6 +136,8 @@ protected function createViewModel($params = null)
             ->fromPost('layout', $this->params()->fromQuery('layout', false));
         if ('lightbox' === $layout) {
             $this->layout()->setTemplate('layout/lightbox');
+        }elseif(isset($params['layout']) && $params['layout']=='simple'){
```
Is there really a need for this 'simple' layout? Is there a reason you can't use 'lightbox'? The only difference seems to be that the lightbox layout includes Piwik/Google Analytics tracking and simple does not. If tracking needs to be disabled for some reason, perhaps there is a way to do that without creating a whole new layout.

bpalme replied:
I might be doing something wrong, but if I just use 'lightbox' as the layout in my action, it still renders the whole page instead of the html snippet that I am using in the getItemstatus JS to show the availability.

demiankatz replied:
The issue is that createViewModel is looking in the HTTP request parameters for the 'lightbox' flag, not in the parameters being sent to the view. For your situation, the best solution is probably to call `$this->layout()->setTemplate('layout/lightbox');` from your controller action instead of trying to pass an extra flag through to createViewModel. Does that work?

bpalme replied:
Do you mean like this?  If so, I'm getting an error: resolver could not resolve to a file...
```PHP     
        $view = $this->createViewModel(
            [
              //'layout' => 'simple',
              'ids' => $ids,
              'action' => $action,
              'result' => $result,
            ]
        );
        //$view->setTemplate('RecordDriver/SolrOverdrive/status-full');
        $this->layout()->setTemplate('layout/lightbox');

        return $view; 
```

demiankatz replied:
Yes, that's correct -- but don't comment out the $view->setTemplate. You'll still need that. One is setting the view template (the "inside" content) and the other is setting the layout template (the "outside" wrapper).

demiankatz commented on module/VuFind/src/VuFind/Controller/OverdriveController.php:
```diff
@@ -0,0 +1,211 @@
+     */
+    public function __construct(ServiceLocatorInterface $sm)
+    {
+        $this->setLogger($sm->get('VuFind\Logger'));
```
You don't need to manually set the logger here -- by virtue of the LoggerAwareInterface being on the class, the initializer code should automatically inject the logger for you.

bpalme replied:
I think I got this from you.  this controller still needs it I guess:
See https://sourceforge.net/p/vufind/mailman/vufind-general/?viewmonth=201803&viewday=13

I just tried commenting it out to see but the logging disappeard.

bpalme replied:
It's dawning on me that I missed a step when I did the PR.  I forgot to get the latest upstream changes and merge them in.  I'll work on that before I continue.

demiankatz replied:
Whoops -- slip of the memory. Sorry about that! :-)

demiankatz commented on module/VuFind/src/VuFind/Controller/OverdriveController.php:
```diff
@@ -0,0 +1,211 @@
+        }
+        
+        $view = $this->createViewModel(
+            [
```
Just a handy shortcut to know about -- when creating arrays like this one, you can use the compact function. In this case, `compact('od_id', 'rec_id', 'action', 'result', 'formats', 'cover', 'title')` will do the job. (Not that you have to do it that way -- but I like shaving off lines of code when I can get away with it.

demiankatz commented on module/VuFind/src/VuFind/Controller/Plugin/Factory.php:
```diff
@@ -210,4 +210,21 @@ public static function getStorageRetrievalRequests(ServiceManager $sm)
+     *
+     * @return StorageRetrievalRequests
+     */
+    public static function getOverdriveConnector(ServiceManager $sm)
```
Is this a factory for something from an earlier design that is no longer needed?

bpalme replied:
I guess it must be.  Will fix in the next commit.

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/Factory.php:
```diff
@@ -0,0 +1,70 @@
+     *
+     * @return OverdriveConnector
+     */
+    public static function getOverdriveConnector(ServiceManager $sm)
```
I'm trying to move away from static factories, since using the new ZF factory interface makes factories more reusable. It's probably worth rewriting this accordingly. See https://github.com/vufind-org/vufind/blob/master/module/VuFind/src/VuFind/Content/AbstractAmazonFactory.php for an example of the new factory style.

bpalme replied:
Ok, just to make sure I understand.  I will create a Vufind\DigitialContent\OverdriveConnectorFactory (or something similar) and implement the _invoke method, right?  Or will it be more generic than that?

demiankatz replied:
Yes, exactly right. One of the main benefits of this new approach is that you create `new $requestedName(...)` instead of using a hard-coded class name. Then, if somebody creates a subclass to customize something, they don't need to copy or extend the default factory unless they change the constructor signature -- one less thing to have to maintain in local custom code.

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -0,0 +1,819 @@
+        // Init session cache for session-specific data
+        if ($conf=$this->getConfig()) {
+            $namespace = md5("VuFind\DigitalContent");
+            $this->sessionContainer = new Container($namespace);
```
When creating a session container, you need to inject the session manager -- I see that you are passing it as a parameter to the constructor, but it doesn't seem to be getting used here.

bpalme replied:
I don't think I needed to inject the session manager at all.  I just instantiate a new session container and start using it.  For now I have removed it from the constructor call.  Let me know if you think I've missed something.

demiankatz replied:
Zend\Session\Container takes a session manager as the second constructor parameter. If you don't inject it, it creates its own -- but that causes problems, because then you end up with two independent, conflicting session managers in use -- the one VuFind created, and the one the container created. This sometimes causes weird, subtle bugs. It's definitely a good idea to inject the manager, though the code will at least superficially appear to work if you do not.

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -0,0 +1,819 @@
+     *
+     * @return string token for the session
+     */
+    private function _callUrl($url)
```
We try to avoid private functions in VuFind, since they limit extensibility. I'd recommend changing this to `protected function callUrl` instead.

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -0,0 +1,819 @@
+        if ($this->_connectToAPI()) {
+            $tokenData = $this->sessionContainer->tokenData;
+            $this->debug("url for OD API Call: $url");
+            $ch = curl_init($url);
```
All of VuFind's HTTP communications are done through the Zend\Http\Client class, and there's a VuFindHttp\HttpService that creates pre-configured Client objects based on config.ini settings. We should rewrite this to utilize the library if possible, for consistency and to make sure that things like proxy server settings are reflected correctly.

bpalme replied:
OK.  I have been considering this but hadn't gotten up the gumption to do it. I also didn't know about the implementation of the HttpService.

demiankatz replied:
Let me know if you'd like any help -- I'd be happy to help convert the code if you're willing to test it, if that would save you time and trouble.

bpalme replied:
So far so good.  I think I'll be able to get it.

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -0,0 +1,819 @@
+ *
+ * @package VuFind\DigitalContent
+ */
+class OverdriveResult
```
This class should probably get its own file for stylistic reasons.

bpalme replied:
I'll probably just use stdClass instead.

demiankatz replied:
Sounds reasonable to me!

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/Factory.php:
```diff
@@ -201,7 +201,33 @@ public static function getSolrMarcRemote(ServiceManager $sm)
+     *
+     * @return SolrOvedrive
+     */
+    public static function getSolrOverdrive(ServiceManager $sm)
```
I recently refactored all of the record driver factories to use the new style; this will have to get moved. Sorry for the nuisance!

bpalme replied:
Should the new factory necessarily extend the SolrMarcFactory if the RecordDriver extends SolrMarc?  the constructor arguments are different since it uses my connector instead of the TitleHolds logic stuff.

demiankatz replied:
No, the factory doesn't have to extend SolrMarcFactory if that doesn't make sense under the circumstances. If it's easier/more logical to make a separate factory, that's fine.

demiankatz commented on themes/bootstrap3/js/check_item_statuses.js:
```diff
@@ -12,13 +12,16 @@ function linkCallnumbers(callnumber, callnumber_handler) {
   return callnumber;
 }
 function displayItemStatus(result, $item) {
+   console.log("displayItemStatus"+result.full_status);
```
I assume you will eventually want to remove these debug statements.

bpalme replied:
Sure I'll clean this up including removing comments and indentation etc pending what we decide about this file in general.

demiankatz commented on themes/bootstrap3/js/check_item_statuses.js:
```diff
@@ -125,8 +124,121 @@ function checkItemStatus(el) {
+*/
 var itemStatusObserver = null;
+
+var ItemStatusHandler = {
```
It looks like you've done some significant reorganization of the item status handling code. Might this be worth separating out into its own pull request, or is it too tangled up with overdrive-specific logic? (I haven't read the code closely enough to tell -- I just know that one of our goals is to contain more of our JS code in objects for better control and clarity, and it seems like that's what you've done here).

bpalme replied:
I can if you prefer.  Most of the changes are to make the code more extendable and then a few lines that actually extend the base object for getting statuses for the overdrive integration.  But either way, it should continue to work as before without Overdrive.

demiankatz replied:
If you don't mind, that would be great! When I find stand-alone chunks in a pull request, I like to separate them out as it makes the overall review process a bit easier, and also lets us gain some benefits to the code that much more quickly.

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/core.phtml:
```diff
@@ -0,0 +1,141 @@
+<div class="media" vocab="http://schema.org/" resource="#record" typeof="<?php echo $this->driver->getSchemaOrgFormats()?> Product">
```
Just food for thought: I wonder if there's a way to implement this without having to duplicate all of the code from DefaultRecord/core.phtml. Again, I haven't done a close comparison, so maybe it's unavoidable to have to do it this way... but if we could achieve the same results by either wrapping extra markup around a render of the default template, or if we could create an optional placeholder variable in DefaultRecord/core that we could pass in from here, that might simplify future code maintenance. This can stand as-is if that's a crazy or impractical idea, though!

demiankatz replied:
(And ditto for all of the other custom record driver templates as well, where applicable).

bpalme replied:
I don't have strong feelings about this.  I did try to keep the templates fairly similar to the default for consistency.  But it could be that users would want to to modify to make them different.  I'm willing to try something else if you want to give me more direction on it.  Otherwise, I'm going to continue down this path for now.

demiankatz replied:
We can stick with this for now and revisit later if it proves to be a maintenance burden.

bpalme replied:
OK, I'm changing my tune on this.  I already have changes coming through on the core template that I didn't realize were there.  Most of the markup I have so far is in one discrete chunk.  Based on that it would seem that your  "optional placeholder" idea would make more sense.  I feel like in older versions of VF you could just include another template.  Is that what you are thinking?  Or maybe a view helper?

demiankatz replied:
Exactly. You can $this->render('path/to/template') to include another template... so you could, hypothetically, do something like:

<pre>
$this->extraContent = "my custom html";
echo $this->render('RecordDriver/DefaultRecord/core.phtml');
</pre>

If the default core.phtml were modified to do something when $this->extraContent is set.

(Contrived example, of course -- but I think you see my idea).

EreMaijala replied:
Now I didn't check the modifications required to the templates, but here's one idea anyway: One option would also be to include predefined extension points in the default templates like this:

    ...
    <?php if ($this->resolver('RecordDriver/DefaultRecord/record-pre-toolbar.phtml')): ?>
      <?=$this->render('RecordDriver/DefaultRecord/record-pre-toolbar.phtml') ?>
    <?php endif; ?>

    <?=$this->record($this->driver)->getToolbar()?>

    <?php if ($this->resolver('RecordDriver/DefaultRecord/record-post-toolbar.phtml')): ?>
      <?=$this->render('RecordDriver/DefaultRecord/record-post-toolbar.phtml') ?>
    <?php endif; ?>
    ...

The three lines above could be put in a helper so that they'd become a single line of template code (like `$this->renderIfExists('...'); ?>`. Now I have no idea if this has performance implications if used a lot, but I suppose it's lighter than always including a set of templates that by default would be empty.

demiankatz replied:
From a design clarity perspective, I definitely prefer @EreMaijala's approach. However, from a performance perspective, my suggestion of injecting content through a variable is probably preferable. In order for the resolver to decide whether or not a template is available, it has to search a significant number of different possible filenames (number of themes in the inheritance chain multipled by number of record drivers in the class hierarchy), which, when combined with multiple extension points and multiple records being displayed in a result set, adds up to a LOT of file system reads. I haven't done benchmarking to see how much this matters in practice, but my gut feeling is that it could be problematic at least under some conditions.

Regardless of whether you take the "pull" or "push" approach to the problem, the most important thing is picking good extension points where content could be inserted. Then it becomes a matter of either overriding the correct templates (if using Ere's way) or setting the correct variables (if using my way). If we design the extension points well, then the nature of the implementation could be changed in the future with relatively little fuss.

bpalme replied:
What about a way to figure out the record type ( Essentially the string that is used to resolve the driver.?)

<? if ($this->record($this->driver)->getType() == "overdrive"): 
      <?= $this->render('RecordDriverOverdrive/Solr/core.phtml'); ?>
<? endif;?>

demiankatz replied:
My goal is to avoid any solution that adds hard-coded logic about specific drivers into code intended to be used generically. That's the sort of thing that's okay for a while but can grow into an unmanageable mess over time. (One special case if statement seems fine... but eventually it can end up being twenty). That's why I prefer a solution that introduces hooks of some sort and allows them to be used by the more specific code.

demiankatz replied:
Note that you always have the option of using `$this->record($this->driver)->renderTemplate($templateName)`, and the code will look for `$templateName` in the template folder specific to your record driver, then look up the inheritance chain until it finds a match. So you can create a hook by building an empty template in the DefaultRecord templates and overriding it in Overdrive. As I mentioned to Ere, I have concerns that doing this kind of thing too much will impact performance... but if that's the easiest way to approach it for now, feel free to try it in this pull request. It will help figure out the logical separations we need, and we can adjust the implementation as needed before we merge to master.

demiankatz replied:
That's strange; I thought that render automatically inherited the current context, so I don't see why that isn't working unless I'm forgetting some fundamental detail. Out of curiosity, does it work differently if you try:

`echo $this->context()->renderInContext('RecordDriver/DefaultRecord/core.phtml', ['OverDriveMarkup => '<p>my markup</p>']);`

?

I don't think you should have to use the context helper for this, but I'd be interested to see if it makes a difference. If nothing else, that might offer a clue.

bpalme replied:
Well I actually had two values to pass through and I was trying to save them in an array.   for some reason, it didn't like that.  So when I changed it to just two different variables it worked.

demiankatz replied:
Hmm, I'm not sure exactly why an array would be a problem... but if two variables works, that's probably fine too. Let me know if you need to dig deeper into this!

demiankatz commented on themes/bootstrap3/templates/myresearch/menu.phtml:
```diff
@@ -52,6 +52,9 @@
       </a>
     <?php endif; ?>
   <?php endif; ?>
+       <a href="<?=$this->url('overdrive-mycontent')?>"<?=$this->active == 'dgcontent' ? ' class="active"' : ''?>>
```
Right now it appears that this menu option is always going to appear -- but presumably we want to hide it when Overdrive is not active.

bpalme replied:
I'm not sure how to handle this one.  I thought maybe there is a way to access the configurations from this template, but I don't see any other template doing that to use as a guide.  Is there a better way to handle this?

demiankatz replied:
There is a Config view helper that you can call from a template with `$this->config()->get('config-name')`, so you could do something like `if ($this->config()->get('Overdrive')->Section->Setting ?? false):` to check a config.

We don't use this view helper very much because many functions have their own purpose-built helpers, and it's often cleaner to put a check method on a function-specific helper than to go rooting around in raw configuration files.... but either approach is possible.

bpalme replied:
I'd like to make this a little smarter by allowing the menu to be hidden when the user doesn't have access to Overdrive.  In my "Overdrive Connector" I have a function that will return this.  Is there a way I can access that without going to to much trouble?

demiankatz replied:
This is probably an argument for creating an Overdrive view helper, since a view helper is the easiest way to expose aspects of a VuFind service to the templates. Basically you'd just need to create a VuFind\View\Helper\Root\Overdrive class extending the base view helper (and a corresponding OverdriveFactory) and register it in the root theme's theme.config.php. The factory could inject the connector into the helper... and then you could create whatever methods you need. Say, for example, you created `Overdrive::isActive()` -- then in a template you could call `$this->overdrive()->isActive()`.

It's a bit of a headache, but probably the best way. If you like, I could probably stub out the classes for you in this branch fairly quickly and let you fill in / test the details.

bpalme replied:
that would be very much appreciated.

demiankatz replied:
See a83d89cb4b57b4aeebd69bf02eb9008f42096304 for an example of how to do this... it might even come close to working, if I guessed correctly what you wanted to do. I'm about out of time for today, but let me know if you have questions/concerns/problems and we can discuss further tomorrow.

I think once this is fully functional, we're going to want to have a bigger conversation about how to name and organize things -- it would probably be better to hide things that reference "Overdrive" behind the more generic "Digital Content" as much as possible to make this more flexible in the long term... but for now I just did it the easy/lazy way so we can get through the "first working prototype" step!

bpalme replied:
Thank you.  this worked just fine.  Are you thinking that there would be a generic DigitalContent view helper?

demiankatz replied:
Exactly. I think the ideal situation would be to develop an abstract interface called DigitalContent that could proxy to Overdrive or other services... the challenge, of course, is that until we have more than one use case, it's hard to figure out what the commonalities are. It may make the most sense to implement direct Overdrive access for now and revisit later if a different use case arises... but if we can figure out a way to make it more abstract up front, then it may be less disruptive to increase flexibility in the future. In any case, once there's a fully functional prototype, it may be easier to decide how to proceed (and I realize we're quite close to that point!).

bpalme commented:
If I want to add just a few css tweaks for the overdrive stuff, where do I do that?  In Bootrap less file?  And i so, do I need to compile that?  We are only talking about a couple of places where i need a little padding for images.  Nothing critical.

demiankatz commented:
@bpalme, yes, as you say, the easiest thing to do is probably to tweak the LESS and recompile the CSS... unless the styles are so specific and used in such a limited context that it would make more sense to create a separate overdrive.css and only include that conditionally on the overdrive pages. Either approach is acceptable -- but the LESS one probably requires the least work.

demiankatz commented:
@bpalme, thanks for the ongoing work here. Just wanted to touch base and see where we were on #1197. Where the comments left off there, you were going to check if something was missing. Did you get a chance to do that? I've been waiting to hear more from you before proceeding to work on that code, but I just want to be sure you're not also waiting for me to do something else. :-)

demiankatz commented on themes/bootstrap3/templates/myresearch/odmycontent.phtml:
```diff
@@ -0,0 +1,152 @@
+  $this->layout()->breadcrumbs = '<li><a href="' . $this->url('myresearch-home') . '">' . $this->transEsc('Your Account') . '</a></li> <li class="active">' . $this->transEsc('od_content') . '</li>';
+?>
+
+<div class="<?php echo $this->layoutClass('mainbody')?>">
```
Just an observation: instead of `<?php echo ... ?>` you can use `<?= ... ?>` to save a bit of typing and improve readability.

bpalme replied:
I thought I saw a comment or something that you wanted to move away from that construction so I've been trying to fix all those in the templates.  I don't know where I saw that.

demiankatz replied:
@bpalme, we moved away from the short tags `<? ... ?>` because it is not recommended anymore (we're using the less ambiguous `<?php ... ?>` instead). However, the short echo tag (`<?= ... ?>`) is still accepted, since it doesn't conflict with tags used by other languages. Sorry for the confusion!

demiankatz commented:
@bpalme, I also just ran php-cs-fixer and phpcbf on the code to reduce the number of outstanding style issues. There's still a bit of manual cleanup that will have to be done before this is ready for merging, but getting the functionality finished is obviously the first priority!

bpalme commented:
All the functionality is there for this version as far as I'm concerned.  I have one or two tiny fixes to make that have come to light.  But I'll add those in and retest.

As for the formats record tab it isn't orphaned;  it has been working fine.  I don't remember how I added it, but I'll double-check it.  It sounds like however I did it, it should be migrated to the plugin manager so I'll do that.

I will also work on the formatting problems.

demiankatz commented:
Great, thanks, @bpalme! Please leave another comment here when you're ready for the next review!

bpalme commented:
I've fixed the few things mentioned above.  I added the formats tab to the plugin manager although it seemed to be working fine with just adding it to module.config.php.  I'll continue to do some more testing to try and be thorough.  But if you want to go ahead and do another review, that would be good.  Do I need to do anything about the CI alert?
also if you want access web to my dev instance with a test card, let me know.

bpalme commented:
I'm done with my testing.  found a few things and committed them.

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -0,0 +1,1557 @@
+    public function getFormatNames()
+    {
+        return [
+            'ebook-kindle' => "Kindle Book",
```
These strings should probably be in a translation file and run through a translator somewhere or else they'll always display in English regardless of language setting.

bpalme replied:
Moved to language file.

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -0,0 +1,1557 @@
+    {
+        $metadata = [];
+        if (!$overDriveIds || count($overDriveIds) < 1) {
+            $this->logWarning("no overdrive content IDs waere passed in.");
```
Typo. :-)

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/SolrOverdrive.php:
```diff
@@ -0,0 +1,662 @@
+    /**
+     * Overdrive Connector
+     *
+     * @var \VuFind\DigitalContent\OverdriveConnector
```
You should add a `use` statement for this class at the top of the file instead of using the fully qualified name here, since you're using the short version in the constructor parameter list already. That would also allow you to add a typehint on the appropriate constructor parameter, which is not a bad idea.

bpalme replied:
I think I did what you were intending here.  I also checked a few other classes for similar issues.

demiankatz commented on module/VuFind/src/VuFind/RecordTab/PluginManager.php:
```diff
@@ -74,6 +75,7 @@ class PluginManager extends \VuFind\ServiceManager\AbstractPluginManager
         CollectionList::class => CollectionListFactory::class,
         Description::class => InvokableFactory::class,
         Excerpt::class => ExcerptFactory::class,
+        Formats::class => InvokableFactory::class,
```
You're using InvokableFactory here, so can you delete the FormatsFactory you defined above? Also, it's probably worth adding an alias `'formats' => Formats::class` for completeness. (There are some defaults that probably auto-guess all of this and make it work, but I try to define all the configs explicitly to clarify what is available).

bpalme replied:
Again, I'm pretty sure I understood this and did it. I removed the factory and added the alias.

demiankatz commented on module/VuFindAdmin/src/VuFindAdmin/Controller/OverdriveController.php:
```diff
@@ -0,0 +1,84 @@
+        $view->productsKey = $connector->getCollectionToken();
+        $view->overdriveConfig = $connector->getConfig();
+        $view->hasAccess = $connector->getAccess();
+        //$view->token = $connector->connectToAPI();
```
Are these commented-out lines significant?

bpalme replied:
removed

demiankatz commented on themes/bootstrap3/templates/RecordDriver/DefaultRecord/core.phtml:
```diff
@@ -1,9 +1,14 @@
     $preview = $this->record($this->driver)->getPreviews();
+
+    //override the default preview link with overdrive link
+    if($this->OverdrivePreview)  $preview = $this->OverdrivePreview;
```
Rather than putting Overdrive-specific language into this file, can we do this more generically? For example, call this `$this->previewOverride` and perhaps replace the `OverdriveAction` and `OverdriveAvail` with a `$this->extraControls` variable that is output if non-empty? That will make this template more reusable with other children without having to add lots of additional special cases.

bpalme replied:
Yes. I like it.

bpalme replied:
I changed my mind on this.  Although the previewOverride kind of makes sense, I can't find a good solution for the "extra controls" that aren't replacing something.  

One of the things I didn't like about our approach for the child template is that it's really hard for someone to override the template in their own theme.  I discovered that when I tried to do it myself.  If it's just on big block of text coming from the child, then you don't have much option for changing the layout.  By at least breaking the two Overdrive blocks (actions buttons and availability block) up, you can move each one around a bit.  If you want to actually change the markup within, it's even more of a pain (but at least you know where to go look for it).  If it's just a generic blob of markup coming from any of the children, how would you know might be in there?  For example, if you know you don't use overdrive, you know you can safely  comment that line out in your custom theme.  If you do and you want to modify the markup, you can try to follow it to the overdrive template and attempt to change it.  

Am I making any sense?  Maybe there is a middle ground...

bpalme replied:
I guess we could do $this->actionControls and $this->availabilityInfo or something

demiankatz replied:
My main issue here is that it's not really helpful to have variables with the name "overdrive" in them, because that leads us down a road where we potentially have many different service-specific exceptions in the core template, which will make it hard to read and maintain. It also doesn't make sense to have tons of specific conditions like this, because we're only ever going to be dealing with one service at a time under normal circumstances. If it's valuable to you to be able to locally override the core template in a way that reflects Overdrive-specific behavior, I believe that you can check `$this->driver->setSourceIdentifier()` to identify when Overdrive is active without having to rely on specific named variables being set.

Also note that if you really want a totally different layout for Overdrive in your local custom version, you can copy the core template from DefaultRecord in the default theme into the Overdrive folder of your custom theme and adjust to your heart's content... then it will be loaded dynamically only when Overdrive record drivers are loaded. This may be more of a headache to maintain in the long term because you'll have to keep it in sync with the core template as that template evolves over time (which is why we generally try to avoid copying and pasting templates), but it's certainly manageable.

Does that address your concern at all, or am I missing the point?

bpalme replied:
Ok so does using the more generic $this->actionControls and $this->availabilityInfo work for you?
or maybe this:
if(!empty($this->extraControls) {
  echo extraControls['actionControls'];
  echo extraControls['availabilityInfo'];
}

demiankatz replied:
Yes, that works for me. I'd maybe include some `?? ''` fallbacks on the ends of those echo statements so notices are not triggered if somebody sets one key but not the other.

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/result-list.phtml:
```diff
@@ -0,0 +1,185 @@
+          
+
+
+      <?php /* We need to find out if we're supposed to display an OpenURL link ($openUrlActive),
```
There's a lot of commented-out code here; is this work in progress or should it be removed?

bpalme replied:
Oh hmm.  I'm wasn't sure what that code does and whether it's applicable to Overdrive Integration or not.  It looks like it's for embedding metadata?  Based on what you know about this, should I try to support this in the record driver?

bpalme replied:
Regarding the Demo connector idea, it does sound like a lot of work.  It might be worth it if is something that could also somehow be used for unit testing.  I don't think I have it in me to do that right now, but if somebody wanted to do it, I could help with providing sample content.  Especially given how little interest there has been.

demiankatz replied:
The OpenURL logic displays links to your link resolver to find full-text content related to the record. I'd think in Overdrive, the need for this would be lessened, but it might still be useful to provide an alternate avenue for obtaining content that is currently checked out.

Also, regarding the Demo connector, I definitely understand your reluctance to build it right now; it definitely seems that it would be a lot of work. Such a connector would definitely be helpful for unit testing, and that's another goal that would be great to meet eventually, but I don't expect you to do all of that work at this stage.

bpalme replied:
Uncommenting the block doesn't seem to break anything for me.  I'll leave it like that for now.  I guess I'm not sure if I need to do anything in my Record Driver to support it if someone configured their system for OpenURLs.

demiankatz replied:
I think it should be okay; that logic should be inherited from the parent class. It's possible that some tweaking might be necessary to identify the types of materials correctly, but we can cross that bridge when we come to it.

demiankatz commented on themes/bootstrap3/templates/admin/menu.phtml:
```diff
@@ -4,4 +4,5 @@
   <a href="<?=$this->url('admin/config')?>" class="list-group-item<?=strtolower($this->layout()->templateName) == "config" ? ' active' : ''?>"><?=$this->transEsc('Configuration')?></a>
   <a href="<?=$this->url('admin/maintenance')?>" class="list-group-item<?=strtolower($this->layout()->templateName) == "maintenance" ? ' active' : ''?>"><?=$this->transEsc('System Maintenance')?></a>
   <a href="<?=$this->url('admin/tags')?>" class="list-group-item<?=strtolower($this->layout()->templateName) == "tags" ? ' active' : ''?>"><?=$this->transEsc('Tag Maintenance')?></a>
+  <a href="<?=$this->url('admin/overdrive')?>" class="list-group-item<?=strtolower($this->layout()->templateName) == "overdrive" ? ' active' : ''?>"><?=$this->transEsc('Overdrive API')?></a>
```
Should we display this conditionally based on whether Overdrive is configured?

bpalme replied:
Yes.  I think you showed me how to do this and I added it.  Maybe it was code that somehow got reverted.  I'll go back through my notes and try to recreate it.

demiankatz replied:
You did something similar in the MyResearch menu; that might be what you're thinking of. Not sure if the same rule applies here or if a different approach is needed.

bpalme replied:
I added another configuration for this menu to Overdrive.ini as I did for the myResearch menu.  There isn't really a global on/off switch for the integration.  If there is a record type that is overdrive then it's on.

demiankatz commented on themes/bootstrap3/templates/footer.phtml:
```diff
@@ -23,6 +23,9 @@
         <li><a href="<?=$this->url('help-home')?>?topic=search&amp;_=<?=time() ?>" data-lightbox class="help-link"><?=$this->transEsc('Search Tips')?></a></li>
         <li><a href="<?=$this->url('content-page', ['page' => 'asklibrary']) ?>"><?=$this->transEsc('Ask a Librarian')?></a></li>
         <li><a href="<?=$this->url('content-page', ['page' => 'faq']) ?>"><?=$this->transEsc('FAQs')?></a></li>
+        <?php if ($this->config()->get('Overdrive')->Overdrive->ODHelpLink ?? false):?>
```
This is another area where it would be nice to come up with more generic language instead of having Overdrive-specific logic in a widely-used template. For example, to display the "Powered by Summon" message required on Summon-related screens, we set `$this->layout()->poweredBy` within the appropriate controller, and there's a placeholder in the layout which makes use of that value. Could we do something similar here?

bpalme replied:
I like the idea of an "extraHelpLinks" array that different components could dynamically add a link to. The footer would add those links to the "help" section of the footer.  

If a user wants to display the overdrive help link, I think it makes sens to display it in the footer of every page, not just the "my overdrive content" page.  What would be a good way to do that?

demiankatz replied:
One possible approach could be to build a HelpLinks view helper, which could have addLink() and renderLinks() methods. Then various components could push links into the helper through addLink(), and it could display all of the collected values through a call to renderLinks() in the footer. Perhaps that's more engineering than is worth the effort for the simple need here, but it's the first solution that comes to mind. I can help build a prototype if that would be useful.

bpalme replied:
Where would the addLink() method get called from in my case?

demiankatz replied:
It could theoretically be called from almost anywhere. If you wanted to globally inject a string, you could call it as part of the helper's factory. If you wanted to inject it in a context-sensitive way, you could call it from templates or controllers where the link is specifically needed.

bpalme replied:
If you think it's useful and want to set up the view helper, then I'll try to use it.  If not, I can just take the link out.  It doesn't matter to me that much.

demiankatz replied:
If you think it's worth the effort, I'm happy to help with the view helper idea... but if you're not that invested in the issue, it's obviously simpler to just remove it for now. We can always revisit the idea later as a separate project.

bpalme replied:
Demian I finally got around to removing the link in the footer.  I believe that I'm caught up on the changes you requested back in March.  Let me know what's next.

demiankatz commented:
@bpalme, I've also run some automatic processes to clean up many of the style issues that Travis was complaining about. Hopefully that will get the build to pass now (or at least significantly reduce the amount of manual work that still needs to be done to comply with style requirements).

bpalme commented on module/VuFind/src/VuFind/Controller/OverdriveController.php:
```diff
@@ -0,0 +1,311 @@
+    {
+        $this->debug("ODC mycontent action");
+
+        //TODO get hold and checkoutlimit using the Patron Info API
```
Just a placeholder for future functionality.  I can remove it if you prefer.

demiankatz replied:
Feel free to leave them if they'll be useful for future reference... but I'd suggest reviewing all of your TODO's throughout the PR in case there are any that are obsolete, just so they don't confuse us in the future.

bpalme replied:
I've reviewed them.  I changed some of the Todos to something else like "future" or something and removed some.

bpalme commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -0,0 +1,1557 @@
+        $this->debug("SolrOverdrive Connector");
+        $this->mainConfig = $mainConfig;
+        $this->recordConfig = $recordConfig;
+        $this->container = $container;
```
I agree and that's the way it was.  The reason I changed it is because we are trying to make more use of the "console" functionality.  ex: a cron job that runs through users' saved searches to find new titles and notify them.  If I have the session controller in there, it chokes when trying to instantiate it in the console context.  Maybe there is a better way to do it...

demiankatz replied:
You can do a check: `if (PHP_SAPI !== 'cli') { ... }` to skip logic that won't work in console mode. Perhaps you can wrap the problem logic in the connector and/or the factory to work around the problem. Does that help?

bpalme replied:
OK I fixed this.  It's working OK for now.  I haven't been able to test it in the console environment yet.  But I can do that later.
also, what is the rationale for not injecting the container?  Does that make it brittle somehow?

bpalme commented:
Thank you very much for helping with the style fixes.  Phpstorm doesn't seem to have configurations for all the intricacies for styling multiple line calls.  In general I feel like the 82 character limit is counter-productive.  It forces line breaks that don't need to be there in my opinion and makes the code less readable.  Is that sort of an industry standard?

demiankatz commented:
@bpalme, thanks for responding to my comments -- I think I've answered all of the conversations that required more input. I've also gotten the Travis build passing. Regarding the line length limit, you can actually stretch up to 85 characters if you need to. This width is something of an industry standard (or at least, it was at the time we instituted the practice, which was quite a few years ago now -- it's possible that norms have adjusted since then). In any case, I find it more helpful than harmful on balance, though I definitely acknowledge that there are cases where it is unavoidably annoying. On the positive side of the spectrum, I like being able to read all of the code without any horizontal scrolling, and I appreciate the discipline that it enforces in terms of not putting too many details and operations on a single line.

bpalme commented on config/vufind/permissions.ini:
```diff
@@ -76,6 +76,9 @@
 ; access.StaffViewTab - Controls access to the staff view tab in record mode
 ; access.SummonExtendedResults - Controls visibility of protected Summon results
 ; feature.Favorites - Controls access to the "save favorites" feature
+[access.AdminModule]
```
This was a mistake.  how does one revert this?  I have tried several variations of this:
`git checkout origin/master config/vufind/permissions.ini`
If you can nuke this from the PR easily, go for it.

demiankatz replied:
In this case, since it's just some added lines, the easiest thing is probably to use the web-based editor and simply delete the offending content. I'll do that right now, which should take care of this issue. Thanks for the clarification!

bpalme commented:
OK, I've been working on these changes.  There are still a couple outstanding that I've commented on.  I'll add comments above as well.  Hopefully I haven't introduced more style problems!

bpalme commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -0,0 +1,1557 @@
+        $container,
+        $ilsAuth
+    ) {
+        $this->debug("SolrOverdrive Connector");
```
removed

bpalme commented:
I ran phpcs on these files so hopefully that will work better.

demiankatz commented:
@bpalme, thanks for the progress! Feel free to ignore the style issues in the PR quality check. For some reason, that external service has started making suggestions that are inconsistent with our project styles. Travis is the authority there. I'll look into whether I can fix the inconsistency!

The other failure seems to be related to lines in en.ini being out of the expected alphabetical order. I've fixed this for you. If it happens again and you want to sort it out, you can fix it automatically from the command line with `php $VUFIND_HOME/public/index.php language normalize $VUFIND_HOME/languages`.

demiankatz commented:
@bpalme, you asked in one of the resolved threads why it is considered bad practice to inject the service container into a service. I can't seem to find the thread again, so I'm just replying here. The idea is that this is bad practice for the same reason that global variables are bad practice -- it makes it harder to understand the relationship between parts of the application, and it's easier to break something in a future refactoring as a result. If a class has an explicit list of dependencies that get injected through the constructor, you know what you need to build that class. If you instead inject the service container, and the class pulls things out of it on demand, then in theory it could depend on any service you have defined. This makes it harder to build tests (since you have to build containers instead of just passing direct dependencies), and it makes things brittle because removing or renaming a service could break a class in a subtle way (e.g. if the service is only fetched down certain specific code paths). And of course that's not to say there aren't reasons why injecting containers can be tempting -- it cuts down function signatures and prevents unnecessary class instantiation in some cases -- but I've found that the maintenance costs outweigh the benefits, and there are other ways to make some of those same gains without container injection.

bpalme commented:
That's a good explanation.  Makes sense to me.

From: Demian Katz [mailto:notifications@github.com]
Sent: Monday, March 04, 2019 9:53 AM
To: vufind-org/vufind <vufind@noreply.github.com>
Cc: Brent Palmer <Brent-Palmer@icpl.org>; Mention <mention@noreply.github.com>
Subject: Re: [vufind-org/vufind] Overdrive integration (#1188)


@bpalme<https://github.com/bpalme>, you asked in one of the resolved threads why it is considered bad practice to inject the service container into a service. I can't seem to find the thread again, so I'm just replying here. The idea is that this is bad practice for the same reason that global variables are bad practice -- it makes it harder to understand the relationship between parts of the application, and it's easier to break something in a future refactoring as a result. If a class has an explicit list of dependencies that get injected through the constructor, you know what you need to build that class. If you instead inject the service container, and the class pulls things out of it on demand, then in theory it could depend on any service you have defined. This makes it harder to build tests (since you have to build containers instead of just passing direct dependencies), and it makes things brittle because removing or renaming a service could break a class in a subtle way (e.g. if the service is only fetched down certain specific code paths). And of course that's not to say there aren't reasons why injecting containers can be tempting -- it cuts down function signatures and prevents unnecessary class instantiation in some cases -- but I've found that the maintenance costs outweigh the benefits, and there are other ways to make some of those same gains without container injection.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/vufind-org/vufind/pull/1188#issuecomment-469303678>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEXrbrHdsYikQTxYL-7klg805O-Qx7KVks5vTUFdgaJpZM4UTsbd>.

demiankatz commented:
Thanks, @bpalme -- I'm going to be at a conference for the rest of this week, but I'll give this another look when I return next week and hopefully get it merged!

demiankatz commented on languages/en.ini:
```diff
@@ -731,6 +731,63 @@ number_decimal_point = "."
+od_none_found = "No titles found."
+od_return_failure = "This title could not be returned."
+od_return_success = "This title has been returned."
+od_video-streaming = "> "streaming video file"
```
@bpalme, there seems to be a weird extra quote here. What is this string supposed to say?

bpalme replied:
That's a copy-paste error.  Should just be:
od_video-streaming = "streaming video file"

demiankatz replied:
Thanks, @bpalme, I've corrected it. I think that means we're completely ready to merge, once you have a chance to test and confirm that my last-minute tweaks didn't break anything.

demiankatz commented:
@bpalme, thanks again for the work on this. I believe that it is ready to merge. However, I made a couple of minor edits that I can't test myself:

1.) Record tab configuration has been moved from module.config.php to RecordTabs.ini to simplify customization... so I had to move the SolrOverdrive tab configuration.

2.) In your most recent changes to allow customization of the core.phtml template, I saw some opportunities to slightly compact/simplify the code, so I made some optimizations.

If you can pull these latest changes and confirm that everything still works correctly for you, then I will merge as soon as I get the all clear from you!

bpalme commented:
I pulled the changes and did a quick run-through of basic functionality.  It was exhaustive but I did the major stuff, submit/cancel hold, check out and return an item etc.  Looks OK to me.

bpalme commented:
Also going forward, if I have little changes to templates or other features, should I just do  pull requests for each of those?

demiankatz commented:
@bpalme, thanks for the double-check. I've gone ahead and merged this. Please feel free to open additional pull requests for minor tweaks and fixes. Those should be able to be merged a lot more quickly than this one was!

Also, I've started a wiki page here to document how this all works -- if you can add some notes there about how you set up indexing, etc., that would be very helpful for future reference:

https://vufind.org/wiki/configuration:overdrive

bpalme commented:
OK, I made some notes in the wiki.  Let me know if you were expecting something different.

demiankatz commented:
@bpalme, this is great, thank you!
